core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -263,7 +263,9 @@ class Partition(val topic: String,
 
   def delete() {
```

Comment:
This function will be called only when `deletePartition` flag is set to true in the StopReplicaRequest and the partition being deleted exists in the `allPartitions` map (See https://github.com/linkedin/kafka/blob/2.0-li/core/src/main/scala/kafka/server/ReplicaManager.scala#L366)

However, during partition reassignment, controller will send out two StopReplicaRequests to the removed replicas in order (See https://github.com/linkedin/kafka/blob/2.0-li/core/src/main/scala/kafka/controller/ControllerChannelManager.scala#L496):
1. _StopReplicaRequest_ (`deletePartition=false`) - in reaction to OfflineReplica state transition
2. _StopReplicaRequest_ (`deletePartition=true`) - in reaction to ReplicaDeletionStarted state transition

When processing the 1st StopReplicaRequest, the broker already removes the partition from `allPartitions` map (See https://github.com/linkedin/kafka/blob/2.0-li/core/src/main/scala/kafka/server/ReplicaManager.scala#L353) so it is guaranteed that `partition.delete()` will not be called during partition reassignment. You can verify it in your test case by sending _StopReplicaRequest_ (`deletePartition=false`) and then _StopReplicaRequest_ (`deletePartition=true`).

I think instead of completing the delated requests in `Partition.scala`, you can complete them in `ReplicaManager.scala` when processing _StopReplicaRequest_.